
# ğŸ§¬ Jumbo Phage Annotation

## combineSPHAEinterpro script

This script will take the results from SPHAE and InterProScan (after running the runinterprobatch.py script) and combine them into a single .gbk output file. It will also generate a summary file and a top_interpro_hits.tsv file for each genome.

This script expects that you have already run SPHAE annotation on all your phages, and that you have a consistent naming prefix for your files (e.g. all files begin with PA- or Phage-). It also expects that you have already run the interprobatch.py script on all your aa.fasta files.  

see: 
- https://github.com/linsalrob/sphae 

for more information. 

requirements: 
- python 3.x

## ğŸš€ Usage

Currently must run from the same directory that combineSPHAEinterpro.py is in (or path to script must be specified in the command). 

You can run the script from the command line as follows:

```bash
python3 combineSPHAEinterpro.py --interpro_folder /path/to/InterproScan_Results/prefix --prefix <prefix> --SPHAE_folder /path/to/SPHAE/final-annotate/ --output_folder /path/to/desired/output/folder/

```

This script expects to be pointed to the example/final-annotate folder generated by SPHAE. This folder should have subdirectories within it for each genome begining with the supplied prefix. The .gbk files are what this script is searching for: 

```bash 
final-annotate/
â”œâ”€â”€ PA-187_Pseudomonas_phage/
â”‚ â””â”€â”€ PA-187_Pseudomonas_phage.gbk
â”œâ”€â”€ PA-329_Pseudomonas_phage/
â”‚ â””â”€â”€ PA-329_Pseudomonas_phage.gbk

```

This script also expects to be pointed to the InterProScan results folder generated by interproscan_batch_script.py, which should have subdirectory within it begining with the supplied prefix. The .tsv files are what this script is searching for. 

for example, if you gave the interproscan_batch.py script the arguments: 

```bash
python3 runinterprobatch.py --prefix PA- --output_folder /path/to/InterproScan_Results/ ...

```

then the argument for this script would be: 
```bash
python3 combineSPHAEinterpro.py --interpro_folder /path/to/InterproScan_Results/PA- --prefix PA- ...

```

file structure for InterProScan results should look like this: 

```bash 
InterproScan_Results/
â””â”€â”€ PA-/
    â”œâ”€â”€ PA-187_Pseudomonas_phage_
    â”‚   â””â”€â”€ PA-187_Pseudomonas_phage_aa.fasta.tsv
    â”œâ”€â”€ PA-329_Pseudomonas_phage_
    â”‚   â””â”€â”€ PA-329_Pseudomonas_phage_aa.fasta.tsv

```

if you run the command without any arguments, you will get a list of required flags: 

```bash
(base) user@MSI:~/folder$ python3 combineSPHAEinterpro.py
usage: combineSPHAEinterpro.py [-h] [--interpro_folder INTERPRO_FOLDER] [--prefix PREFIX]
                               [--SPHAE_folder SPHAE_FOLDER] [--output_folder OUTPUT_FOLDER]

Combine InterProScan results with SPHAE results in a GBK file, with summary.txt and top_interpro_hits.tsv files generated as well.

options:
  -h, --help            show this help message and exit
  --interpro_folder INTERPRO_FOLDER
                        Path to the input folder containing subdirectories with interproscan .tsv files.
  --prefix PREFIX       Prefix to filter the aa.fasta files (e.g. PA-, KA-, Phage-).
  --SPHAE_folder SPHAE_FOLDER
                        Path to the SPHAE results folder (usually labeled 'final-annotate/' by SPHAE).
  --output_folder OUTPUT_FOLDER
                        Path to the output folder where results will be saved.

```

## ğŸ“ Output

Output will be a .gbk, .txt, and .tsv file for each genome. Results will show up in a "<prefix>_results_combined" subdirectory in the specified output folder.

the .gbk file will have a new qualifier in each CDS feature that had interproscan hits. this qualifier is called "/interpro_product" and will contain the top interproscan hit for that CDS.

If the CDS was labeled as hypothetical by SPHAE but interproscan found a hit, the existing product qualifier will be appended to with " | <interpro_product>". The source qualifier will also be updated to include "InterProScan" for all interproscan hits. 

Example snippet of .gbk file: 
```bash

     CDS             36479..38517
                     /ID="KACWZWFB_CDS_0061"
                     /transl_table=11
                     /phrog="6574"
                     /locus_tag="KACWZWFB_CDS_0061"
                     /function="unknown function"
                     /product="hypothetical protein | Phage Pam3 gp32-like
                     protein"
                     /interpro_product="Phage Pam3 gp32-like protein"
                     /source="PHANOTATE_1.6.7, InterProScan"
                     /translation="MSRVFVVQNQHRWDRDKQRFEPKFNLAPAEEFGELVYLLSPTAAP
                     FRPEPIIDELKEKLADFGPGDHLLLVGNPVLIGFAVAIAADANDGDVSLLQWSGKDQRY
                     IAVDASGLFTSSD"
                     /protein_id="PA-187_Pseudomonas_phage_60"
     CDS             38572..39145
                     /ID="KACWZWFB_CDS_0062"
                     /transl_table=11
                     /phrog="8239"
                     /locus_tag="KACWZWFB_CDS_0062"
                     /function="DNA, RNA and nucleotide metabolism"
                     /product="HNH endonuclease"
                     /interpro_product="Pam3 gp33-like protein"
                     /source="PHANOTATE_1.6.7, InterProScan"
                     /translation="MSSENAYLDYVQPQTAGGELSQLTQLAEQQAAAQLKVSDLEAQLN
                     KAREELRDIAERQVPELMDQIGIGEFKTTTGLKIKIDETIRASIPKAKAPLAFAWLKQN
                     GHAAMIKRVVAVAFGKGEDEKADELRQRLADEFEVDDNASVHPSTLAAFVREKLREGEE
                     VPLDLFGVHRQRVSKIEV"
                     /protein_id="PA-187_Pseudomonas_phage_61"

```

The .tsv file will contain the top interproscan hit for each CDS, with the following columns: 
```bash
Locus	Product	E-value	Source
KACWZWFB_CDS_0004	all-alpha NTP pyrophosphatases	9.16E-13	InterproScan
KACWZWFB_CDS_0014	3'-5' exoribonuclease Rv2179c-like domain	3.9E-27	InterproScan
KACWZWFB_CDS_0017	GGDEF	7.05827E-24	InterproScan
KACWZWFB_CDS_0018	Antirestriction protein	4.3E-43	InterproScan

```

the summary file will contain a summary of the interproscan hits for each genome, with the following information: 
```bash 
Summary for PA-187_Pseudomonas_phage:
Total CDS: 90
Annotated by InterProScan: 32
Annotated by SPHAE: 50
Annotated by just InterProScan (NEW): 5
Annotated by just SPHAE: 23
Annotated by both InterProScan and SPHAE: 27
Fully hypothetical (no SPHAE or Interpro hits): 35

Newly annotated CDS:
	KACWZWFB_CDS_0025 -> YuA phage, Gp49
	KACWZWFB_CDS_0061 -> Phage Pam3 gp32-like protein
	KACWZWFB_CDS_0069 -> Domain of unknown function (DUF7300)
	KACWZWFB_CDS_0070 -> Amino acid:DNA transferase
	KACWZWFB_CDS_0075 -> Nucleotide modification associated domain 5

```


# ğŸ™‹â€â™€ï¸ Author/ ğŸ“¬ Contact

For questions of suggestions, contact: 

Hannah Kapoor
ğŸ“§ hannahkapoor00@gmail.com 
