# 🧬 Jumbo Phage Annotation

## runinterprobatch script

This script will take all the predicted amino acid sequences by SPHAE and run them through interproscan to detect any interdomain annotations that might have been missed. 

This script expects that you have already run SPHAE annotation on all your phages, and that you have a consistent naming prefix for your files (e.g. all files begin with PA- or Phage-). It also expects that you have interproscan downloaded locally and can run interproscan.sh from the command line. 

see: 
- https://github.com/linsalrob/sphae 
- https://interproscan-docs.readthedocs.io/en/v5/HowToDownload.html

for more information. 


### requirements: 
- python 3.x
- interproscan 

## 🚀 Usage

Currently must run from the same directory that runinterprobatch.py is in (or path to script must be specified in the command). 

You can run the script from the command line as follows:

```bash
python3 runinterprobatch.py --input /path/to/example/PROCESSING/genome-annotate/ --prefix <prefix> --output_folder /path/to/output/folder --interpro_path /path/to/interproscan.sh 

```

This script expects to be pointed to the PROCESSING/genome-annotate folder generated by SPHAE. This folder should have subdirectories within it, with -pharokka, -phold, -phynteny, and -predict directories for each genome. The -predict folders are what this script is searching for. within each predict folder, there should be a aa.fasta file. 

```bash 
PROCESSING/
└── genome-annotate/
    ├── PA-187_Pseudomonas_phage-pharokka/
    ├── PA-187_Pseudomonas_phage-phold/
    ├── PA-187_Pseudomonas_phage-phynteny/
    ├── PA-187_Pseudomonas_phage-predict/
    │   └── PA-187_Pseudomonas_phage-predict_aa.fasta
    ├── PA-329_Pseudomonas_phage-pharokka/
    ├── PA-329_Pseudomonas_phage-phold/
    ├── PA-329_Pseudomonas_phage-phynteny/
    ├── PA-329_Pseudomonas_phage-predict/
    │   └── PA-329_Pseudomonas_phage-predict_aa.fasta

```

if you run the command without any arguments, you will get a list of required flags: 

```bash
(base) user@MSI:~/Folder$ python3 runinterprobatch.py
usage: runinterprobatch.py [-h] [--input_folder INPUT_FOLDER] [--prefix PREFIX] [--output_folder OUTPUT_FOLDER]
                           [--interpro_path INTERPRO_PATH]

Run interproscan on all phage aa.fasta files of a given species from SPHAE output.

options:
  -h, --help            show this help message and exit
  --input_folder INPUT_FOLDER
                        Path to the input folder containing subdirectories with aa.fasta files (usually PROCESSING/genome-annotate in SPHAE
                        output).
  --prefix PREFIX       Prefix to filter the aa.fasta files (e.g. PA-, KA-, Phage-).
  --output_folder OUTPUT_FOLDER
                        Path to the output folder where results will be saved.
  --interpro_path INTERPRO_PATH
                        path to the interproscan.sh script
```

## 📝 Output

Output will be interproscan results for each genome in the given input directory, located in the given output directory. Each genome will have it's own folder with the prefix and naming conventions provided at the command line, with an aa.fasta.gff3, aa.fasta JSON, aa.fasta.tsv, and aa.fasta HTML file for each genome annotated. 

Here is an example aa.fasta.tsv output for a phage genome: 
```bash
_PA-319_Pseudomonas_phage:GXRZAZWF_CDS_0213	9533aa3e7f92ef62e2ad9599e5f2b1de	571	Pfam	PF07230	Bacteriophage T4-like portal protein (Gp20)	49	508	5.1E-99	T	05-08-2025	IPR010823	Portal protein Gp20	-	-
_PA-319_Pseudomonas_phage:GXRZAZWF_CDS_0337	b5377956a7fa8a73f5d7f82c04112218	576	SMART	SM00382	AAA_5	189	325	2.1E-22	T	05-08-2025	IPR003593	AAA+ ATPase domain	-	-
_PA-319_Pseudomonas_phage:GXRZAZWF_CDS_0337	b5377956a7fa8a73f5d7f82c04112218	576	FunFam	G3DSA:1.10.8.60:FF:000001	ATP-dependent zinc metalloprotease FtsH	324	398	9.7E-26	T	05-08-2025	-	-	-	-
_PA-319_Pseudomonas_phage:GXRZAZWF_CDS_0338	b5377956a7fa8a73f5d7f82c04112218	576	FunFam	G3DSA:3.40.50.300:FF:000277	ATP-dependent zinc metalloprotease FtsH	137	322	2.4E-85	T	05-08-2025	-	-	-	-
```

# 🙋‍♀️ Author/ 📬 Contact

For questions or suggestions, contact: 

Hannah Kapoor
📧 hannahkapoor00@gmail.com 
